commands/cluster-setslot.md

```diff
@@ -79,3 +79,14 @@ Notes:
   If the source node is informed before the destination node and the destination node crashes before it is set as new slot owner, the slot is left with no owner, even after a successful failover.
 * Step 6, sending `SETSLOT` to the nodes not involved in the resharding, is not technically necessary since the configuration will eventually propagate itself.
   However, it is a good idea to do so in order to stop nodes from pointing to the wrong node for the hash slot moved as soon as possible, resulting in less redirections to find the right node.
+* Starting from Valkey 8.0, `CLUSTER SETSLOT` is synchronously replicated to all healthy replicas
```

Comment:
I don't want to put synchronous replication anywhere in the docs. It's not technically synchronous since we don't revert it on failures, so it's only one copy of the data. I would prefer saying "is replicated to all healthy replicas before being executed on the primaries.

Reply:
>since we don't revert it on failures, so it's only one copy of the data.

I don't think this is the common definition of  "synchronous" replication. I modeled our "synchronous" replication after [postgress'](since we don't revert it on failures, so it's only one copy of the data.) 

> I would prefer saying "is replicated to all healthy replicas before being executed on the primaries.

This is also not accurate nor complete because we do wait for a certain number of replicas to ack the replication and this is the core part of this feature offering.

Reply:
I guess "reconfiguration" includes adding and removing nodes. Resharding is only about moving slots?

I'm fine with either of the terms. :)

Reply:
> I don't think this is the common definition of "synchronous" replication. I modeled our "synchronous" replication after [postgress'](https://www.postgresql.org/docs/current/warm-standby.html#SYNCHRONOUS-REPLICATION). The "revert" part falls in the realm of "distributed transactions" and it is a step up from "synchronous replication".

In postgreSQL, everything is a transaction though. There is no "step up" so to speak, they are the same.

> This is also not accurate nor complete because we do wait for a certain number of replicas to ack the replication and this is the core part of this feature offering.

"Is replicated and acknlowedged on all healthy replicas before being executed on the primary"?

Reply:
>In postgreSQL, everything is a transaction though.

I am not aware that Postgress' replication is done transactionally based on my research. My understanding is that there is no rollback on other replicas when one replica fails to ack the replication either. It is just that the primary will wait for replicas to ack the replication. Note that in order to achieve transactional replication, one would need to implement 2PC, which trades (more) performance/availability for consistency. Do you have a Postgress code or doc pointer that indicates their repliction is done transactionally?  I can dig some more too.

>"Is replicated and acknlowedged on all healthy replicas before being executed on the primary"?

Yes. 

The "synchronous" qualifier is critical here and it is used to contrast the current "asynchronous" replication, which doesn't provide the required semantics. 

BTW, not related to Postgress, but "synchronous replication" is also used in the exact same semantics as I used here for another database I worked on before.

Reply:
I vote for explaining this without using the word "synchronous". It's controversial to introduce this term.

Reply:
>However, we don't guarantee synchronous replication since the replicas can be down at the time, which is my bigger concern at the moment

I am convinced by your argument that this is not a strict synchronous replication so "sync repl" is indeed overselling. However, I still think we need a name for this concept to differentiate from the normal replication behavior today.

I think the concept we implemented is closer to MySQL's [semi-synchronous replication](https://dev.mysql.com/doc/refman/8.4/en/replication-semisync.html) but even that offers more guarantee. I wonder if "Opportunistic Synchronous Replication" would capture our idea more accurately?

>Semisynchronous replication falls between asynchronous and fully synchronous replication. The source waits until at least one replica has received and logged the events (the required number of replicas is configurable), and then commits the transaction. The source does not wait for all replicas to acknowledge receipt, and it requires only an acknowledgement from the replicas, not that the events have been fully executed and committed on the replica side. Semisynchronous replication therefore guarantees that if the source crashes, all the transactions that it has committed have been transmitted to at least one replica.

Reply:
I'm not sure anymore. We actually do wait for "all healthy replicas" before committing, so this seems synchronous to me. If we timeout, we don't rollback this from the replicas that already ACKed this though.

Reply:
I'd still vote for no name. I asked a bunch of folks at AWS. The most common response was "don't invent a name" if it's not an industry standard, since that just adds more confusion to this mix. "Best effort synchronous replication" came up from a few folks, which I don't really like. Someone suggested "replicated with acknowledgement", which I kind of like but not sure if it resonates as a "good name".

Reply:
"Best effort" == "Opportunistic" :) I think we will need a name at some point IF we expand this pattern beyond this command but I am on board with not adding a new concept for this one. I will drop "synchronous" but keep the high level explanation in this doc.

Reply:
Best effort is kind of the key concept since the very beginning. Memcached doesn't have persistence and replication. Redis/Valkey has it, best effort, async, without compromising latency.

The cluster bus is supposed to be truly consistent though. We'll fix that some day.

Reply:
Btw, the MIGRATE command doesn't use the replication nor cluster bus. It just sets up a new connection. Is it synchronous? It waits for OK before deleting the transferred key, I think, or does it? Is this how we should have done SETSLOT?

Reply:
> "Best effort" == "Opportunistic"

Yes. I would be OK with either of those two FWIW, it's not my top preference.

Reply:
I dropped "synchronous". PTAL